File: src/main/java/com/github/jaksonlin/pitestintellij/actions/RunCaseAnnoationCheckAction.java
package com.github.jaksonlin.pitestintellij.actions;

import com.github.jaksonlin.pitestintellij.commands.unittestannotations.CheckAnnotationCommand;
import com.github.jaksonlin.pitestintellij.context.CaseCheckContext;
import com.github.jaksonlin.pitestintellij.util.Pair;
import com.github.jaksonlin.pitestintellij.util.PsiUtil;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiMethod;
import org.jetbrains.annotations.NotNull;


public class RunCaseAnnoationCheckAction extends AnAction {

    @Override
    public void actionPerformed(@NotNull AnActionEvent e) {
        Pair<PsiMethod, PsiClass> psiMethodInfo = PsiUtil.findMethodAtCaret(e);
        if (psiMethodInfo == null) {
            return;
        }
        CaseCheckContext context = CaseCheckContext.create(psiMethodInfo.getKey(), psiMethodInfo.getValue());
        Project project = e.getProject();
        if (project != null) {
            CheckAnnotationCommand command = new CheckAnnotationCommand(project, context);
            command.execute();
        }
    }

}

File: src/main/kotlin/com/github/jaksonlin/pitestintellij/actions/RunCaseAnnoationCheckAction.kt
package com.github.jaksonlin.pitestintellij.actions

import com.github.jaksonlin.pitestintellij.commands.unittestannotations.CheckAnnotationCommand
import com.github.jaksonlin.pitestintellij.context.CaseCheckContext
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiDocumentManager
import com.intellij.psi.PsiMethod
import com.intellij.psi.util.PsiTreeUtil

class RunCaseAnnoationCheckAction : AnAction() {

    override fun actionPerformed(e: AnActionEvent) {
        val psiMethodInfo = findMethodAtCaret(e) ?: return
        val context = CaseCheckContext.create(psiMethodInfo.first, psiMethodInfo.second)
        CheckAnnotationCommand(e.project!!, context).execute()
    }

    private fun findMethodAtCaret(e: AnActionEvent): Pair<PsiMethod, PsiClass>? {
        val project = e.project ?: return null
        val editor = e.dataContext.getData(CommonDataKeys.EDITOR) ?: return null
        val caret = e.dataContext.getData(CommonDataKeys.CARET) ?: return null
        val elementAtCaret = PsiDocumentManager.getInstance(project)
            .getPsiFile(editor.document)?.findElementAt(caret.offset) ?: return null
        val method = PsiTreeUtil.getParentOfType(elementAtCaret, PsiMethod::class.java) ?: return null
        val containingClass = PsiTreeUtil.getParentOfType(method, PsiClass::class.java) ?: return null
        return method to containingClass
    }

}